stdlib/LinearAlgebra/src/tridiag.jl

```diff
@@ -405,7 +405,7 @@ struct Tridiagonal{T,V<:AbstractVector{T}} <: AbstractMatrix{T}
     dl::V    # sub-diagonal
     d::V     # diagonal
     du::V    # sup-diagonal
-    du2::V   # supsup-diagonal for pivoting in LU
+    du2::Union{Nothing, V}   # supsup-diagonal for pivoting in LU
```

Comment:
Why is this needed?

Reply:
If we keep `du2::V` then non `isbittype` containers will contain junk values. This causes issues when converting between types.

For example (in this patch, if we don't have that)
```
julia> x = Tridiagonal(1:2, 1:3, 1:2)
3×3 Tridiagonal{Int64,UnitRange{Int64}}:
 1  1  ⋅
 1  2  2
 ⋅  2  3

julia> x.du2
0:140331565583360
```
if we convert this to type `Tridiagonal{Int64, Vector{Int64}}` it will try to convert `du2` and we get an `OutOfMemoryError`.

By letting it be `Nothing` we don't have to worry about the case where `undef` gives junk values.

Reply:
I missed https://github.com/JuliaLang/julia/pull/23154 (had just gotten my second son) and therefore the changes to `du2`. The original (but not necessarily the right/better) assumption with including `du2` in the `Tridiagonal` struct was that you'd always do `\` with it, i.e. the main application of a `Tridiagonal` is to solve a linear system of equations. Hence it would be fine so allocate `du2` upfront. If we decided to stick with that assumption then we could probably just always initialize `du2` to something meaningful and avoid the `Union` here. If we believe that `Tridiagonal`s are generally used beyond linear solves, I think introducing a `Union` here probably makes sense.

Reply:
Assuming general usage seems like the better choice to me, but I defer to you guys. https://github.com/JuliaLang/julia/pull/23154/files#r131596874 brings up an important thing that I have been thinking about/discussing about range types with these matrices. The simple case of `D = Diagonal(1:3); D-D` fails because of the range type. But that isn't immediately relevant to this PR.

I'll wait to see where the discussion about du2 leads before updating that part.

Reply:
Right. You can only to the with `StepRangeLen` (the float default) but not for `UnitRange`.

Reply:
Should we just ignore that case and let it fail? That should be documented somewhere probably.